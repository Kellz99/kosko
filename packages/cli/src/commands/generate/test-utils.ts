import { Manifest } from "@kosko/generate";

export function createManifest(
  value: Partial<Omit<Manifest, "report">> = {}
): Manifest {
  return {
    position: { path: "", index: [] },
    issues: [],
    data: undefined,
    report() {},
    ...value
  };
}
