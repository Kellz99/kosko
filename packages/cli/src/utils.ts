export function excludeFalsyInArray<T>(input: (T | undefined | null)[]): T[] {
  return input.filter(Boolean) as T[];
}
