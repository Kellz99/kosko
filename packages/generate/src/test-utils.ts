/// <reference types="jest-extended" />
import { Manifest } from "./base";

type ManifestToMatch = Pick<Manifest, "data" | "metadata"> &
  Partial<Pick<Manifest, "position" | "issues">>;

export function matchManifest({
  position = { path: "", index: [] },
  issues = [],
  ...manifest
}: ManifestToMatch) {
  return {
    ...manifest,
    position,
    issues,
    report: expect.toBeFunction()
  };
}

export function matchManifests(manifests: ManifestToMatch[]) {
  return manifests.map(matchManifest);
}
