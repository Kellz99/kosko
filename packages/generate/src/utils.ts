export function validateConcurrency(value = 10): number {
  if (value < 1) {
    throw new Error("Concurrency must be greater than 0");
  }

  return value;
}
