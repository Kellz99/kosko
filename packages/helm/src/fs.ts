import { getErrorCode } from "@kosko/common-utils";
import { cp, rename, rm, stat } from "node:fs/promises";

export async function move(src: string, dest: string): Promise<void> {
  try {
    await rename(src, dest);
  } catch (err) {
    if (getErrorCode(err) !== "EXDEV") throw err;

    await cp(src, dest, {
      recursive: true,
      errorOnExist: true,
      preserveTimestamps: true
    });
    await rm(src, { recursive: true, force: true });
  }
}

export async function fileExists(path: string): Promise<boolean> {
  try {
    const stats = await stat(path);
    return stats.isFile();
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;
    return false;
  }
}
