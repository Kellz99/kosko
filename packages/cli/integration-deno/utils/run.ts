import execa from "execa";
import { join } from "node:path";
import { projectRoot } from "@kosko/test-utils";

export async function runDenoCLI(
  args: string[],
  options: execa.Options = {}
): Promise<execa.ExecaReturnValue> {
  return execa(
    "deno",
    [
      "run",
      "-A",
      "--import-map",
      join(projectRoot, "packages/cli/templates/deno/import_map.json"),
      "npm:kosko",
      ...args
    ],
    {
      ...options,
      env: {
        LC_ALL: "en_US",
        ...options.env
      }
    }
  );
}
