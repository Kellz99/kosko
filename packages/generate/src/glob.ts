import { readdir } from "fs/promises";
import { join, posix } from "path";
import mm from "micromatch";

function createMatcher(patterns: readonly string[], options?: mm.Options) {
  const matchers = patterns.map((pattern) => mm.matcher(pattern, options));
  return (path: string): boolean => matchers.some((matcher) => matcher(path));
}

export interface GlobOptions {
  path: string;
  patterns: readonly string[];
  extensions: readonly string[];
}

export interface GlobResult {
  relativePath: string;
  absolutePath: string;
}

export async function* glob(options: GlobOptions): AsyncIterable<GlobResult> {
  const matcherOptions: mm.Options = { cwd: options.path };
  const suffix =
    options.extensions.length > 1
      ? `.{${options.extensions.join(",")}}`
      : `.${options.extensions.join(",")}`;
  const fileMatcher = createMatcher(
    options.patterns.map((pattern) => pattern + suffix),
    matcherOptions
  );
  const dirMatcher = createMatcher(options.patterns, matcherOptions);

  async function* walk(
    path: string,
    parent: string
  ): AsyncIterable<GlobResult> {
    for (const file of await readdir(path, { withFileTypes: true })) {
      const result: GlobResult = {
        relativePath: posix.join(parent, file.name),
        absolutePath: join(path, file.name)
      };

      if (file.isDirectory()) {
        if (dirMatcher(result.relativePath)) {
          yield result;
        }

        yield* walk(result.absolutePath, result.relativePath);
      } else if (fileMatcher(result.relativePath)) {
        yield result;
      }
    }
  }

  yield* walk(options.path, "");
}
