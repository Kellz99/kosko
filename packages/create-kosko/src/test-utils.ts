import { join, posix } from "node:path";
import glob from "fast-glob";
import { readFile } from "node:fs/promises";

export async function listAllFiles(
  dir: string
): Promise<Record<string, string>> {
  const paths = await glob("**/*", { cwd: dir });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(join(dir, path), "utf8");
  }

  return files;
}
