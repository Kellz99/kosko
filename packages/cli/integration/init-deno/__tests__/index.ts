import { makeTempDir, TempDir } from "@kosko/test-utils";
import glob from "fast-glob";
import { join, posix } from "node:path";
import { readFile } from "node:fs/promises";
import { runDenoCLI } from "../../utils/run";

let tmpDir: TempDir;

beforeEach(async () => {
  tmpDir = await makeTempDir();
});

afterEach(async () => {
  await tmpDir.cleanup();
});

test("should setup a new folder", async () => {
  await runDenoCLI(["init", tmpDir.path]);

  const paths = await glob("**/*", { cwd: tmpDir.path });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(
      join(tmpDir.path, path),
      "utf8"
    );
  }

  expect(files).toMatchSnapshot();
});
