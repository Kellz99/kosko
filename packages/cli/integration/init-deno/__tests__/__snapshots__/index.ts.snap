// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print usage guide to console 1`] = `
"info  - Creating a Kosko project in "<INIT_PATH>"
info  - Project created at "<INIT_PATH>"

Inside that directory, you can run several commands:

  deno task kosko generate
    Validate components and generate Kubernetes manifests.

  deno task kosko validate
    Only validate components.

We suggest that you begin by typing:

  cd <INIT_PATH>
  deno task kosko generate"
`;

exports[`should setup a new folder 1`] = `
{
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
deno task kosko generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
deno task kosko generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
deno task kosko generate --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.ts": "import { Deployment } from "npm:kubernetes-models/apps/v1/Deployment";
import { Service } from "npm:kubernetes-models/v1/Service";
import env from "npm:@kosko/env";

const params = await env.component("nginx");

const name = "nginx";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: "nginx",
            image: "nginx:stable",
            ports: [{ containerPort: port }],
            env: [{ name: "NGINX_PORT", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

export default [deployment, service];
",
  "deno.json": "{
  "compilerOptions": {
    "types": ["./environments/types.d.ts"]
  },
  "tasks": {
    "kosko": "deno run -A npm:kosko/deno.js"
  }
}
",
  "environments/dev/index.ts": "import { GlobalEnvironment } from "npm:@kosko/env";

const env: GlobalEnvironment = {
  namespace: "dev"
};

export default env;
",
  "environments/dev/nginx.ts": "import { ComponentEnvironment } from "npm:@kosko/env";

const env: ComponentEnvironment["nginx"] = {
  replicas: 3
};

export default env;
",
  "environments/types.d.ts": "import "npm:@kosko/env";

// Extend type declarations of "@kosko/env" module
declare module "npm:@kosko/env" {
  // Declare types for global environment variables
  interface GlobalEnvironment {
    namespace: string;
  }

  // Declare types for component environment variables
  interface ComponentEnvironment {
    // Fallback type of all other component variables which are not specified below
    [key: string]: unknown;

    // Specify types for each component
    nginx: {
      replicas: number;
    };
  }

  // Extend Environment interface
  interface Environment {
    global(): Promise<GlobalEnvironment>;

    component<K extends string>(
      name: K
    ): Promise<GlobalEnvironment & ComponentEnvironment[K]>;
  }
}
",
  "kosko.toml": "components = ["*"]
",
}
`;
